### DISCUSSION TITLE
Post mortem analysis of dead data

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#112] [Design & Development]
We need some standard functions like the simple print_dead_logs I posted in the pull request for the bundle code.  The general need is tools to address two questions a user will always want to know:

1.  How many errors were generated by function x, 
2. What were the errors logged by function x applied to a dataset (rdd or dask bag)?

The problem is more complex than the serial world, but the clear solution would use the elog entries automatically (now) saved to MongoDB.   Question 1 would be commonly useful to ask during a workflow.   I can imagine aborting a workflow if the fraction of data deleted exceeded some threshold.   Mostly, however, I can imagine a user would naturally want to do the following at the completion of a data processing run:
* for each processing step tell me how many objects failed?
* For those that had errors write them to a file I can read or copy them into an editors buffer directly.

**Comment by pavlis** [comment]
Working on the docs again and realized two more holes related to dead data:
1.  We need a standard processing function that can be run in parallel that will (a) handle dead data with the Undertaker bury_the_dead mechanism, and (b) delete the dead data from an RDD or dask bag.    I have only a rough idea how that would be implemented so I just punt it to one of you.  There will be a hole in the doc in a doc I'm about done with called "handling_errors.rst" that covers dead data along with a set of other related topics on errors.  
2. For some reason I don't understand the Undertaker class is not in our current sphynx documentation even though the file Undertaker.py is in the repository as mspasspy.util.Undertaker.   Why I don't understand is don't know a lot about how the whole sphynx doc generation is supposed to work.

**Reply by pavlis** [reply]
ps:  looking at the implementation the existing methods need an obvious extension.   bury_the_dead needs to accept atomic data.  The other methods in Undertaker make sense only for ensembles, but bury_the_dead is our mechanism to save dead data error logs.

**Comment by wangyinz** [comment]
I tend to push back this topic a little bit since I think it is sort of independent and doesn't relate to our current database API implementation. Also, this topic is somewhat related to our yet to be implemented global-history mechanism. It will be better to come back to this topic when we have a good grasp of our global history design.

**Reply by pavlis** [reply]
Fine.  I would urge you, however, to start assembling something that lists the things we punt down the road lik this.   I insert this kind of thing because I know from experience that if I don't do so the idea will evaporate.  Maybe we should start a development to-do list with links to these pages.

**Reply by wangyinz** [reply]
This is what the "Project" thing in GitHub is for. But maybe this topic is not a well-defined "project" at the moment. I guess I will just insert this discussion to our design doc for global history.
